### PR TITLE
MiKo_2038 is now aware of some more phrases

### PR DESCRIPTION
**Pull request by RalfKoban**
- Expand set of recognized problematic documentation phrases

- Refactor phrase generation to be more programmatic

- Add a cleanup step to handle edge cases

- Updated Verbalizer to properly make infinite verbs from gerund verbs where the verb ends with `ort` (such as `export`)

### CODE REVIEW

File: MiKo.Analyzer.Tests/Rules/Documentation/MiKo_2038_CommandTypeSummaryAnalyzerTests.cs [hunk 1]
```diff
@@ -114,33 +125,78 @@ public class TestMe : ICommand
...
+        [TestCase("Interface for a toggle command in", "be used in")]
+        [TestCase("Interface for a toggle command that can do", "do")]
+        [TestCase("Interface for a toggle command which can do", "do")]
+        [TestCase("Interface for an async command that can do", "do")]
+        [TestCase("Interface for an command in", "be used in")]
+        [TestCase("Interface for an command that can do", "do")]
+        [TestCase("Interface for an command which can do", "do")]
+        [TestCase("Interface for an toggle command in", "be used in")]
+        [TestCase("Interface for an toggle command that can do", "do")]
+        [TestCase("Interface for an toggle command which can do", "do")]
```
[the remark this answers is not in the document]
Comment (RalfKoban) [reply MiKo.Analyzer.Tests/Rules/Documentation/MiKo_2038_CommandTypeSummaryAnalyzerTests.cs]: typo intended as users might have made that typo

File: MiKo.Analyzer.Shared/Rules/Documentation/MiKo_2038_CodeFixProvider.cs [hunk 2]
```diff
@@ -15,28 +16,22 @@ public sealed class MiKo_2038_CodeFixProvider : SummaryDocumentationCodeFixProvi
...
 
+        private static readonly Pair[] CleanupMap =
+                                                    {
+                                                        new Pair(" commands for ", " "),
+                                                        new Pair(" command for ", " "),
+                                                        new Pair(" can in ", " can be used in "),
+                                                        new Pair(" can with ", " can be used with "),
+                                                    };
+
+        private static readonly string[] CleanupMapKeys = CommandReplacementMap.ToArray(_ => _.Key);
```
[the remark this answers is not in the document]
Comment (RalfKoban) [reply MiKo.Analyzer.Shared/Rules/Documentation/MiKo_2038_CodeFixProvider.cs]: Fixed

File: MiKo.Analyzer.Tests/Rules/Documentation/MiKo_2038_CommandTypeSummaryAnalyzerTests.cs [hunk 3]
```diff
@@ -114,33 +125,78 @@ public class TestMe : ICommand
...
-        [TestCase("This class tries to execute something.", "Represents a command that can execute something.")]
-        [TestCase("Tries to execute something.", "Represents a command that can execute something.")]
+        [TestCase("A class that offers to execute", "execute")]
+        [TestCase("A class that tries to execute", "execute")]
+        [TestCase("A command executes", "execute")]
+        [TestCase("A command that can do", "do")]
+        [TestCase("A interface for commands in", "be used in")]
+        [TestCase("A interface for toggle commands in", "be used in")]
+        [TestCase("A standard command to execute", "execute")]
+        [TestCase("A sync command to execute", "execute")]
```
[the remark this answers is not in the document]
Comment (RalfKoban) [reply MiKo.Analyzer.Tests/Rules/Documentation/MiKo_2038_CommandTypeSummaryAnalyzerTests.cs]: typo intended as users have made that typo

File: MiKo.Analyzer.Shared/Linguistics/Verbalizer.cs [hunk 4]
```diff
@@ -22,7 +22,7 @@ internal static class Verbalizer
 
         private static readonly string[] FourCharacterGerundEndings = { "pping", "rring", "tting" };
 
-        private static readonly string[] ThreeCharacterGerundEndings = { "anging", "inging", "ssing", "cting", "pting", "enting" };
+        private static readonly string[] ThreeCharacterGerundEndings = { "anging", "inging", "ssing", "cting", "pting", "rting", "enting", };
```
[the remark this answers is not in the document]
Comment (RalfKoban) [reply MiKo.Analyzer.Shared/Linguistics/Verbalizer.cs]: Fixed.

File: MiKo.Analyzer.Shared/Rules/Documentation/MiKo_2038_CodeFixProvider.cs [hunk 5]
```diff
@@ -59,32 +68,148 @@ internal static SyntaxNode GetUpdatedSyntax(SyntaxNode syntax)
...
+        private static HashSet<string> CreateCommandStartingPhrases()
+        {
+            var result = new HashSet<string>();
+
+            var phrases = CreateCommandStartingPhrasesLocal();
+
+            foreach (var phrase in phrases)
+            {
+                result.Add(phrase.ToUpperCaseAt(0));
+                result.Add(phrase.ToLowerCaseAt(0));
```
[the remark this answers is not in the document]
Comment (RalfKoban) [reply MiKo.Analyzer.Shared/Rules/Documentation/MiKo_2038_CodeFixProvider.cs]: Nope. Case sensitivity is needed here to have proper replacements.

File: MiKo.Analyzer.Tests/Rules/Documentation/MiKo_2038_CommandTypeSummaryAnalyzerTests.cs [hunk 6]
```diff
@@ -114,33 +125,112 @@ public class TestMe : ICommand
...
+        [TestCase("A class that offers to execute", "execute")]
+        [TestCase("A class that tries to execute", "execute")]
+        [TestCase("A command base class that tries to execute", "execute")]
+        [TestCase("A command executes", "execute")]
+        [TestCase("A command that can do", "do")]
+        [TestCase("A helper command that is able to do something", "do something")]
+        [TestCase("A helper command to be able to do something", "do something")]
+        [TestCase("A helper command which is able to do something", "do something")]
+        [TestCase("A interface for commands in", "be used in")]
+        [TestCase("A interface for toggle commands in", "be used in")]
```
[the remark this answers is not in the document]
Comment (RalfKoban) [reply MiKo.Analyzer.Tests/Rules/Documentation/MiKo_2038_CommandTypeSummaryAnalyzerTests.cs]: typo intended as users make those and we have to be aware of them

File: MiKo.Analyzer.Tests/Rules/Documentation/MiKo_2038_CommandTypeSummaryAnalyzerTests.cs [hunk 7]
```diff
@@ -114,33 +125,112 @@ public class TestMe : ICommand
...
+        [TestCase("A standard command to execute", "execute")]
+        [TestCase("A sync command to execute", "execute")]
+        [TestCase("A synchronous command to execute", "execute")]
+        [TestCase("A toggle command to execute", "execute")]
+        [TestCase("An async command to execute", "execute")]
+        [TestCase("An asynchronous command to execute", "execute")]
+        [TestCase("An command base class that tries to execute", "execute")]
```
[the remark this answers is not in the document]
Comment (RalfKoban) [reply MiKo.Analyzer.Tests/Rules/Documentation/MiKo_2038_CommandTypeSummaryAnalyzerTests.cs]: typo intended as users make those and we have to be aware of them

File: MiKo.Analyzer.Shared/Rules/Documentation/MiKo_2038_CodeFixProvider.cs [hunk 8]
```diff
@@ -15,28 +15,35 @@ public sealed class MiKo_2038_CodeFixProvider : SummaryDocumentationCodeFixProvi
...
-                                                                      "This toggle command ",
-                                                                      "Command ",
-                                                                      "command ",
-                                                                      "A class ",
-                                                                      "The class ",
-                                                                      "This class ",
-                                                                  };
-
-        private static readonly Pair[] CommandReplacementMap = CreateCommandReplacementMapEntries().OrderDescendingByLengthAndText(_ => _.Key);
+        private static readonly string[] CommandStartingPhrases = CreateCommandStartingPhrases().Except(Constants.Comments.CommandSummaryStartingPhrase).ToArray();
```
[the remark this answers is not in the document]
Comment (RalfKoban) [reply MiKo.Analyzer.Shared/Rules/Documentation/MiKo_2038_CodeFixProvider.cs]: This is wrong. There is a method `Except` within this repository that's doing as expected.

File: MiKo.Analyzer.Shared/Rules/Documentation/MiKo_2038_CodeFixProvider.cs [hunk 9]
```diff
@@ -59,32 +68,148 @@ internal static SyntaxNode GetUpdatedSyntax(SyntaxNode syntax)
...
+                                      "is capable for",
+                                      "capable to",
+                                      "capable for",
+                                      "able to",
                                       "offers to",
                                       "tries to",
                                   };
 
-            var results = new List<Pair>();
+            var results = new HashSet<Pair>();
```
[the remark this answers is not in the document]
Comment (RalfKoban) [reply MiKo.Analyzer.Shared/Rules/Documentation/MiKo_2038_CodeFixProvider.cs]: The type `Pair` contains overrides for `Equals`/`GetHashCode` and is already widely used in `HashSet`s for that exact purpose (prevent duplicates).

File: MiKo.Analyzer.Shared/Rules/Documentation/MiKo_2038_CodeFixProvider.cs [hunk 10]
```diff
@@ -59,32 +68,148 @@ internal static SyntaxNode GetUpdatedSyntax(SyntaxNode syntax)
...
+                        yield return "The interface for " + article + adverb + "command ";
+                        yield return "Interface for " + article + adverb + "command ";
+
+                        yield return "Represents " + article + adverb + "command ";
+                    }
+                }
+
+                yield return "Provides functionality ";
+                yield return "Provides a functionality ";
+                yield return "Provides an functionality ";
```
[the remark this answers is not in the document]
Comment (RalfKoban) [reply MiKo.Analyzer.Shared/Rules/Documentation/MiKo_2038_CodeFixProvider.cs]: Typo intended as users make these kinds of errors

File: MiKo.Analyzer.Shared/Rules/Documentation/MiKo_2038_CodeFixProvider.cs [hunk 11]
```diff
@@ -59,32 +68,148 @@ internal static SyntaxNode GetUpdatedSyntax(SyntaxNode syntax)
...
+                yield return "Provides the functionality ";
+
+                yield return "A interface ";
+                yield return "An interface ";
```
[the remark this answers is not in the document]
Comment (RalfKoban) [reply MiKo.Analyzer.Shared/Rules/Documentation/MiKo_2038_CodeFixProvider.cs]: Typo intended as the users make these kinds of errors.
